src/Generators/ModelGenerator.php

```diff
@@ -190,51 +190,64 @@ protected function buildRelationships(Model $model)
 
         foreach ($model->relationships() as $type => $references) {
             foreach ($references as $reference) {
+                // if reference starts with \ , we assume developer is using a fully namespaced model
+                $is_model_fqn = Str::startsWith($reference, '\\');
+
                 $custom_template = $template;
                 $key = null;
                 $class = null;
 
                 $column_name = $reference;
-                $method_name = Str::beforeLast($reference, '_id');
+                $method_name = $is_model_fqn ? Str::afterLast($reference, '\\') : Str::beforeLast($reference, '_id');
 
                 if (Str::contains($reference, ':')) {
                     [$foreign_reference, $column_name] = explode(':', $reference);
-                    $method_name = Str::beforeLast($column_name, '_id');
+
+                    $method_name = $is_model_fqn ? Str::afterLast($foreign_reference, '\\') : Str::beforeLast($column_name, '_id');
 
                     if (Str::contains($foreign_reference, '.')) {
                         [$class, $key] = explode('.', $foreign_reference);
 
                         if ($key === 'id') {
                             $key = null;
                         } else {
-                            $method_name = Str::lower($class);
+                            $method_name = $is_model_fqn ? Str::lower(Str::afterLast($class, '\\')) : Str::lower($class);
                         }
                     } else {
                         $class = $foreign_reference;
                     }
                 }
 
-                $class_name = Str::studly($class ?? $method_name);
-                $fqcn = $this->fullyQualifyModelReference($class_name) ?? $model->fullyQualifiedNamespace() . '\\' . $class_name;
+                // if full model namespace is proviced we will not try to infer it,
+                // we use the namespace as developer gives us
+                if ($is_model_fqn) {
+                    $fqcn = $class ?? $column_name;
+                    $class_name = Str::afterLast($fqcn, '\\');
+                }else{
```

Comment:
```suggestion
                } else {
```

Reply:
@tabacitu I don't think this was submitted correctly...